src/Raven.Server/Documents/Handlers/BulkInsertHandler.cs

```diff
@@ -6,28 +6,64 @@
 
 using System;
 using System.Collections.Concurrent;
-using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
 using System.Net.WebSockets;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
-using Raven.Abstractions;
 using Raven.Abstractions.Data;
 using Raven.Abstractions.Logging;
+using Raven.Abstractions.Util;
 using Raven.Server.Routing;
 using Raven.Server.ServerWide.Context;
 using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Voron;
+using Voron.Exceptions;
 using Bits = Sparrow.Binary.Bits;
 
 namespace Raven.Server.Documents.Handlers
 {
     public class BulkInsertHandler : DatabaseRequestHandler
     {
+        private const int NumberOfBuffersUsed = 8;
         private readonly BlockingCollection<BulkBufferInfo> _fullBuffers = new BlockingCollection<BulkBufferInfo>();
         private readonly BlockingCollection<BulkBufferInfo> _freeBuffers = new BlockingCollection<BulkBufferInfo>();
 
+        public static readonly ArraySegment<byte> ProcessingMessage = new ArraySegment<byte>(Encoding.UTF8.GetBytes("{'Type': 'Processing'}"));
+        public static readonly ArraySegment<byte> WaitingMessage = new ArraySegment<byte>(Encoding.UTF8.GetBytes("{'Type': 'Waiting'}"));
+        public static readonly ArraySegment<byte> CompletedMessage = new ArraySegment<byte>(Encoding.UTF8.GetBytes("{'Type': 'Completed'}"));
+        public static byte[] ProcessedMessageArray = new byte[1024];
+        public static readonly ArraySegment<byte> ProcessedMessage = new ArraySegment<byte>(ProcessedMessageArray);
+
+        private readonly BlockingCollection<ArraySegment<byte>> _messagesQueue = new BlockingCollection<ArraySegment<byte>>();
+        private int _numbrtOfMsgs;
+
+        private static object lockObj = new object();
+        private static int x;
+
+
+        private void SendMessageToClient(ArraySegment<byte> msg)
+        {
+            // make sure only one thread performs the send async
+            var num = Interlocked.Increment(ref _numbrtOfMsgs);
+            _messagesQueue.Add(msg);
+            ArraySegment<byte> msgToSend;
+            if (num > 1)
+                return;
+            while (num > 0)
+            {
+                if (_messagesQueue.TryTake(out msgToSend))
+                {
+                    AsyncHelpers.RunSync(() => _webSocket.SendAsync(msg, WebSocketMessageType.Text, true, Database.DatabaseShutdown));
+                    num = Interlocked.Decrement(ref _numbrtOfMsgs);
+                    if (num == 0)
+                        break;
+                }
+            }
+        }
```

Comment:
This instead of simple : lock (obj) { SendAsync }.    However it doesn't solve "Unexcpect reserved bits set" problem